### PR TITLE
[GraphBolt] Add opt to load tasks selectively.

### PR DESCRIPTION
**Pull request by yxy235**
## Description

Enable load tasks selectively.
**Example code:** 
```
dataset.load(["node_classification"])
```
The type of the option should be `list`.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
@yxy235 I'm afraid you misunderstood. what user want is to optionally load task instead of train/val/test set in task. just select on **task-level** as we could define any number of tasks in single `metadata.yaml` of dataset.

**Comment by Rhett-Ying** [comment]
for example, we defined 2 tasks: node classification, link prediction in https://docs.dgl.ai/en/latest/stochastic_training/ondisk_dataset_homograph.html#Generate-tasks. And we want to load one of them only. As for the train/val/test set, we usually need to load all of them for the target task.

**Comment by yxy235** [comment]
> for example, we defined 2 tasks: node classification, link prediction in https://docs.dgl.ai/en/latest/stochastic_training/ondisk_dataset_homograph.html#Generate-tasks. And we want to load one of them only. As for the train/val/test set, we usually need to load all of them for the target task.

OK. I will rewrite the code.

**Comment by yxy235** [comment]
@dgl-bot

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 1]
```diff
@@ -415,14 +415,20 @@ def _convert_yaml_path_to_absolute_path(self):
                                 self._dataset_dir, data["path"]
                             )
 
-    def load(self):
+    def load(self, selected_task: list = None):
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:418]:
```suggestion
    def load(self, tasks: List[str] = None):
```

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 2]
```diff
@@ -415,14 +415,20 @@ def _convert_yaml_path_to_absolute_path(self):
...
         """Load the dataset."""
+        if selected_task and not isinstance(selected_task, list):
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:420]: move such sanity check into `self._init_tasks()`

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 3]
```diff
@@ -415,14 +415,20 @@ def _convert_yaml_path_to_absolute_path(self):
                                 self._dataset_dir, data["path"]
                             )
 
-    def load(self):
+    def load(self, selected_task: list = None):
         """Load the dataset."""
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:419]: add detailed docstring including parameters so that user could figure out how to load task selectively.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 4]
```diff
@@ -462,20 +468,26 @@ def all_nodes_set(self) -> Union[ItemSet, ItemSetDict]:
         self._check_loaded()
         return self._all_nodes_set
 
-    def _init_tasks(self, tasks: List[OnDiskTaskData]) -> List[OnDiskTask]:
+    def _init_tasks(
+        self, tasks: List[OnDiskTaskData], selected_task: list
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:472]:
```suggestion
        self, tasks: List[OnDiskTaskData], selected_tasks: List[str]
```

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 3 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:419]: For example, `we're loading via task name .....`

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 5]
```diff
@@ -462,20 +468,26 @@ def all_nodes_set(self) -> Union[ItemSet, ItemSetDict]:
...
         for task in tasks:
-            ret.append(
-                OnDiskTask(
-                    task.extra_fields,
-                    self._init_tvt_set(task.train_set),
-                    self._init_tvt_set(task.validation_set),
-                    self._init_tvt_set(task.test_set),
+            if (
+                selected_task == ["all"]
+                or task.extra_fields["name"] in selected_task
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:481]: as `name` is not a required field for now, accessing it blindly may throw exception here.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 6]
```diff
@@ -462,20 +468,26 @@ def all_nodes_set(self) -> Union[ItemSet, ItemSetDict]:
...
             return ret
         for task in tasks:
-            ret.append(
-                OnDiskTask(
-                    task.extra_fields,
-                    self._init_tvt_set(task.train_set),
-                    self._init_tvt_set(task.validation_set),
-                    self._init_tvt_set(task.test_set),
+            if (
+                selected_task == ["all"]
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:480]: no need to add `all` key word, just check if it's `None`.

File: tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py [hunk 7]
```diff
@@ -2223,3 +2223,54 @@ def test_OnDiskTask_repr_heterogeneous():
...
+        yaml_file = os.path.join(test_dir, "metadata.yaml")
+        with open(yaml_file, "w") as f:
+            f.write(yaml_content)
+
+        # Case1. Test load all tasks.
+        dataset = gb.OnDiskDataset(test_dir).load()
+        assert len(dataset.tasks) == 2
+
+        # Case2. Test load tasks selectively.
+        dataset = gb.OnDiskDataset(test_dir).load(["link_prediction"])
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py:2264]: for kwargs, always specify the key word which helps reader about the meaning of passed argument.

File: tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py [hunk 8]
```diff
@@ -2223,3 +2223,54 @@ def test_OnDiskTask_repr_heterogeneous():
...
+        assert len(dataset.tasks) == 1
+        assert dataset.tasks[0].metadata["name"] == "link_prediction"
+
+        # Case3. Test load tasks with non-existent task name.
+        dataset = gb.OnDiskDataset(test_dir).load(["fake-name"])
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py:2269]: for non-existent tasks, an exception or warning is preferable.

File: tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py [hunk 9]
```diff
@@ -2223,3 +2223,54 @@ def test_OnDiskTask_repr_heterogeneous():
...
+        assert len(dataset.tasks) == 0
+
+        # Case4. Test load tasks selectively with incorrect task type.
+        with pytest.raises(TypeError):
+            dataset = gb.OnDiskDataset(test_dir).load("link_prediction")
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py:2274]: This should not thrown exception. for single task, str should be acceptable. we could convert it to list of str internally.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 10]
```diff
@@ -462,20 +503,39 @@ def all_nodes_set(self) -> Union[ItemSet, ItemSetDict]:
...
+                        self._init_tvt_set(task.validation_set),
+                        self._init_tvt_set(task.test_set),
+                    )
+                )
+                if selected_tasks:
+                    task_names.add(task_name)
+        if selected_tasks:
+            fake_task_name = set(selected_tasks) - task_names
+            if len(fake_task_name):
+                warnings.warn(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:536]: usually `dgl_warning()` is used.

File: tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py [hunk 11]
```diff
@@ -2223,3 +2223,58 @@ def test_OnDiskTask_repr_heterogeneous():
...
+        # Case2. Test load tasks selectively.
+        dataset = gb.OnDiskDataset(test_dir).load(tasks="link_prediction")
+        assert len(dataset.tasks) == 1
+        assert dataset.tasks[0].metadata["name"] == "link_prediction"
+        dataset = gb.OnDiskDataset(test_dir).load(tasks=["link_prediction"])
+        assert len(dataset.tasks) == 1
+        assert dataset.tasks[0].metadata["name"] == "link_prediction"
+
+        # Case3. Test load tasks with non-existent task name.
+        with pytest.warns(UserWarning):
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_ondisk_dataset.py:2272]: match the warning message too.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 12]
```diff
@@ -462,20 +503,39 @@ def all_nodes_set(self) -> Union[ItemSet, ItemSetDict]:
...
+                        task.extra_fields,
+                        self._init_tvt_set(task.train_set),
+                        self._init_tvt_set(task.validation_set),
+                        self._init_tvt_set(task.test_set),
+                    )
+                )
+                if selected_tasks:
+                    task_names.add(task_name)
+        if selected_tasks:
+            fake_task_name = set(selected_tasks) - task_names
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:534]: `not_found_tasks`

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 13]
```diff
@@ -462,20 +503,39 @@ def all_nodes_set(self) -> Union[ItemSet, ItemSetDict]:
...
+            if len(fake_task_name):
+                warnings.warn(
+                    f"The task name {fake_task_name} may not be feasible."
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:537]:
```suggestion
                    f"Below tasks are not found in YAML: {not_found_tasks}. Skipped."
```

**Review by Rhett-Ying — APPROVED** [review]
